src/libmongoc/src/mongoc/mongoc-generation-map.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-generation-map-private.h"
+
+#include "utlist.h"
+
+typedef struct _gm_node_t {
```

Comment:
This is a simple map implemented as a singly-linked list. I err'ed on simplicity instead of performance for the generation map. This map will have at most one entry per unique backing server behind a load balancer.

Reply:
I've never seen a C map implemented as a singly-linked list. I do wonder about performance (particularly in `mongoc_generation_map_get/increment`). Is there precedence for map implementations in the C driver? Any thoughts @chardan?

Reply:
@benjirewis So, on one hand an associative array can be implemented any number of ways, which are all going to have different performance characteristics. @kevinAlbs has noted that this isn't intended for performance, with the key being that it will "have at most one entry per unique backing server". If this number is nearly always small, I see no real problem with using a linked list. 

I did have another couple questions, though, which is that it seems odd we'd have to do this afresh (although this is C land, not C++ land); does the library we're using (is it libcoap..?) for the LL_* macros maybe has something for associative arrays already? My other question appears on line 52, which is that it seems unusual to wrap the data and the link together inside another type, but it could be that @kevinAlbs is intending to make that whole implementation abstract? If so, it almost feels like it should be its own mini-library?

Reply:
+1 as per this and our offline discussioln, sounds entirely fine to me. Not only is N small enough, we have a centralized and internal way to change the implementation! :-)

Reply:
Thanks for the explanations of performance and implementation here. All sounds good to me.